File: integration-tests/pkg/cluster/go/validation_test.go
package cluster

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/require"
)

func fixturePath(name string) string {
	return filepath.Join("..", "fixtures", name)
}

func TestValidateFixtures(t *testing.T) {
	cases := []struct {
		name     string
		file     string
		wantErr  bool
		contains string
	}{
		{name: "valid", file: "load-valid.yaml"},
		{name: "minimal", file: "load-minimal.yaml"},
		{name: "networks present in range", file: "validation-networks-present.yaml"},
		{name: "anchors allowed", file: "validation-anchors.yaml"},
		{name: "missing name", file: "validation-missing-name.yaml", wantErr: true, contains: "name"},
		{name: "missing machines", file: "validation-missing-machines.yaml", wantErr: true, contains: "machines"},
		{name: "empty machines", file: "validation-empty-machines.yaml", wantErr: true, contains: "machines"},
		{name: "missing id", file: "validation-missing-id.yaml", wantErr: true, contains: "missing properties: 'id'"},
		{name: "missing type", file: "validation-missing-type.yaml", wantErr: true, contains: "missing properties: 'type'"},
		{name: "missing platform", file: "validation-missing-platform.yaml", wantErr: true, contains: "platform"},
		{name: "missing private ip", file: "validation-missing-private-ip.yaml", wantErr: true, contains: "privateIP"},
		{name: "unsupported platform", file: "validation-unsupported-platform.yaml", wantErr: true, contains: "platform"},
		{name: "ip outside", file: "validation-ip-outside.yaml", wantErr: true, contains: "must be inside"},
		{name: "missing networks", file: "validation-missing-networks.yaml", wantErr: true, contains: "when 'usePrivateNetwork' is true"},
		{name: "single network", file: "validation-single-network.yaml", wantErr: true, contains: "when 'usePrivateNetwork' is true"},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			data, err := os.ReadFile(fixturePath(tt.file))
			require.NoError(t, err)

			raw, err := parseToMap(data)
			require.NoError(t, err)

			err = validateCluster(raw)
			if tt.wantErr {
				require.Error(t, err)
				if tt.contains != "" {
					require.Contains(t, err.Error(), tt.contains)
				}
				return
			}
			require.NoError(t, err)
		})
	}
}

func TestDefaultsApplied(t *testing.T) {
	data, err := os.ReadFile(fixturePath("load-defaults.yaml"))
	require.NoError(t, err)

	raw, err := parseToMap(data)
	require.NoError(t, err)

	err = validateCluster(raw)
	require.NoError(t, err)

	kubeVersion, err := schemaDefault("properties", "kubernetesVersion", "default")
	require.NoError(t, err)
	talosImage, err := schemaDefault("properties", "talosImage", "default")
	require.NoError(t, err)
	serverType, err := schemaDefault("properties", "machineDefaults", "properties", "hcloud", "properties", "serverType", "default")
	require.NoError(t, err)
	datacenter, err := schemaDefault("properties", "machineDefaults", "properties", "hcloud", "properties", "datacenter", "default")
	require.NoError(t, err)

	require.Equal(t, kubeVersion, raw["kubernetesVersion"])
	require.Equal(t, talosImage, raw["talosImage"])

	machines, ok := raw["machines"].([]any)
	require.True(t, ok)
	require.Len(t, machines, 1)

	machine := machines[0].(map[string]any)
	hcloud := machine["hcloud"].(map[string]any)
	require.Equal(t, serverType, hcloud["serverType"])
	require.Equal(t, datacenter, hcloud["datacenter"])
}

File: integration-tests/pkg/cluster/python/validation_test.py
from __future__ import annotations

from pathlib import Path
from typing import Optional
import json

import pytest
import yaml
from cluster.python.validation import validate_cluster
from cluster.python.defaults import get_default

FIXTURES_DIR = Path(__file__).resolve().parents[1] / "fixtures"
SCHEMA_PATH = Path(__file__).resolve().parents[1] / "schema.json"

with open(SCHEMA_PATH, "r", encoding="utf-8") as f:
    _SCHEMA = json.load(f)

@pytest.mark.parametrize(
    ("fixture_name", "message"),
    [
        ("load-valid.yaml", None),
        ("load-minimal.yaml", None),
        ("validation-networks-present.yaml", None),
        ("validation-anchors.yaml", None),
        ("validation-missing-name.yaml", "Invalid cluster spec: 'name' is a required string"),
        ("validation-missing-machines.yaml", "Invalid cluster spec: 'machines' must be a non-empty array"),
        ("validation-empty-machines.yaml", "Invalid cluster spec: 'machines' must be a non-empty array"),
        ("validation-missing-id.yaml", "Invalid cluster spec: 'machines[0].id' is a required string"),
        ("validation-missing-type.yaml", "Invalid cluster spec: 'machines[0].type' is a required string"),
        ("validation-missing-platform.yaml", "Invalid cluster spec: 'machines[0].platform' is a required string"),
        ("validation-unsupported-platform.yaml", "Invalid cluster spec: 'machines[0].platform' must be 'hcloud'"),
        ("validation-ip-outside.yaml", "Invalid cluster spec: machine 'worker-1' privateIP '10.0.1.10' must be inside '10.0.0.0/24'"),
        ("validation-missing-networks.yaml", "when 'usePrivateNetwork' is true, both 'privateNetwork' and 'privateSubnetwork' are required"),
        ("validation-single-network.yaml", "when 'usePrivateNetwork' is true, both 'privateNetwork' and 'privateSubnetwork' are required"),
        ("validation-unknown-top.yaml", "unknown field 'extra' is not allowed"),
        ("validation-unknown-machine.yaml", "unknown field 'machines[0].unknown' is not allowed"),
    ],
)
def test_validate_cluster(fixture_name: str, message: Optional[str]) -> None:
    data = _load_fixture(fixture_name)
    if message is None:
        validate_cluster(data)
        return

    with pytest.raises(ValueError) as exc:
        validate_cluster(data)
    assert message in str(exc.value)

def _load_fixture(name: str) -> dict:
    with open(FIXTURES_DIR / name, "r", encoding="utf-8") as f:
        return yaml.safe_load(f) or {}


def test_defaults_are_applied() -> None:
    data = _load_fixture("load-defaults.yaml")

    validate_cluster(data)

    kubernetes_version = get_default(["properties", "kubernetesVersion", "default"])
    talos_image = get_default(["properties", "machines", "items", "properties", "talosImage", "default"])

    machine = data["machines"][0]
    server_type = get_default(
        ["properties", "machineDefaults", "properties", "hcloud", "properties", "serverType", "default"]
    )
    datacenter = get_default(
        ["properties", "machineDefaults", "properties", "hcloud", "properties", "datacenter", "default"]
    )

    assert data["kubernetesVersion"] == kubernetes_version
    assert machine["talosImage"] == talos_image
    assert machine["hcloud"]["serverType"] == server_type
    assert machine["hcloud"]["datacenter"] == datacenter
